consensus/e2e_tests/utils_test.go

```diff
@@ -526,40 +537,221 @@ func baseRpcMock(t *testing.T, _ modules.EventsChannel) *mockModules.MockRPCModu
+	clck *clock.Mock,
+	eventsChannel modules.EventsChannel,
+	pocketNodes IdToNodeMapping,
+	proposalsToSend []*anypb.Any,
+	height uint64,
+	step uint8,
+	round uint8,
+	leaderId typesCons.NodeId,
+	noOfExpectedMsgs int,
+	maxWaitTime time.Duration,
+	failOnExtraMessages bool,
+) ([]*anypb.Any, error) {
+
+	for _, message := range proposalsToSend {
+		P2PBroadcast(t, pocketNodes, message)
+	}
+	advanceTime(t, clck, 10*time.Millisecond)
+
+	proposalMsgs, err := WaitForNetworkConsensusEvents(t, clck, eventsChannel, typesCons.HotstuffStep(step), consensus.Propose, noOfExpectedMsgs, maxWaitTime, failOnExtraMessages)
+	if err != nil {
+		return nil, err
+	}
+
+	for nodeId, pocketNode := range pocketNodes {
+		nodeState := GetConsensusNodeState(pocketNode)
+		if (typesCons.HotstuffStep(step) == consensus.Decide) && (nodeId == leaderId) {
+			assertNodeConsensusView(t, nodeId,
+				typesCons.ConsensusNodeState{
+					Height: height + 1,
+					Step:   1,
+					Round:  round,
+				},
+				nodeState)
+			require.Equal(t, typesCons.NodeId(0), nodeState.LeaderId, "Leader should be empty")
+			continue
+		} else {
+			assertNodeConsensusView(t, nodeId,
+				typesCons.ConsensusNodeState{
+					Height: height,
+					Step:   step,
+					Round:  round,
+				},
+				nodeState)
+			require.Equal(t, leaderId, nodeState.LeaderId, fmt.Sprintf("%d should be the current leader", leaderId))
 		}
-	}).AnyTimes()
+	}
 
-	return stateMachineMock
+	return proposalMsgs, nil
+}
+
+func waitForVotes(t *testing.T,
+	clck *clock.Mock,
+	eventsChannel modules.EventsChannel,
+	pocketNodes IdToNodeMapping,
+	prepareProposal []*anypb.Any,
+	noOfExpectedMsgs int,
+	step int,
+	maxWaitTime time.Duration,
+	failOnExtraMessages bool,
+) ([]*anypb.Any, error) {
+
+	for _, message := range prepareProposal {
+		P2PBroadcast(t, pocketNodes, message)
+	}
+	advanceTime(t, clck, 10*time.Millisecond)
+
+	votes, err := WaitForNetworkConsensusEvents(t, clck, eventsChannel, typesCons.HotstuffStep(step), consensus.Vote, noOfExpectedMsgs, maxWaitTime, failOnExtraMessages)
+
+	return votes, err
+}
+
+// waitForNodeToSync waits for a node to sync to a target height
+// For every missing block for the unsynced node:
+//
+//	first, waits for the node to request a missing block via `waitForNodeToRequestMissingBlock()` function,
+//	then, waits for the node to receive the missing block via `waitForNodeToReceiveMissingBlock()` function,
+//	finally, wait for the node to catch up to the target height via `waitForNodeToCatchUp()` function.
+func waitForNodeToSync(t *testing.T,
+	clck *clock.Mock,
+	eventsChannel modules.EventsChannel,
+	unsyncedNode *shared.Node,
+	allNodes IdToNodeMapping,
+	targetHeight uint64,
+) error {
+
+	currentHeight := unsyncedNode.GetBus().GetConsensusModule().CurrentHeight()
+
+	for i := currentHeight; i <= targetHeight; i++ {
+		blockRequest, err := waitForNodeToRequestMissingBlock(t, clck, eventsChannel, allNodes, currentHeight, targetHeight)
+		if err != nil {
```

Comment:
Question: Why did you decide to return the error instead of just `require.NoError(t, err)` in the tests?

Reply:
I made it return err, to make `waitForVotes()` and `waitForProposalMsgs()` to return error to the caller, and caller for example `waitForNextBlock()` executes `require.NoError(t, err)`. I thought this makes error handling more clear, your previous comment on why `waitForNodeToSync()` does not return error, motivated me for this. 

But I don't have strong reasons to have it one way or another.  Please let me know which version is the best.